Tools/clinic/clinic.py

```diff
@@ -5419,14 +5446,20 @@ def update_function_kind(self, fullname: str) -> None:
         _, cls = self.clinic._module_and_class(fields)
         if name in unsupported_special_methods:
             fail(f"{name!r} is a special method and cannot be converted to Argument Clinic!")
+
         if name == '__new__':
-            if (self.kind is not CLASS_METHOD) or (not cls):
+            if (self.kind is CLASS_METHOD) and cls:
+                self.kind = METHOD_NEW
+            else:
                 fail("'__new__' must be a class method!")
-            self.kind = METHOD_NEW
         elif name == '__init__':
-            if (self.kind is not CALLABLE) or (not cls):
-                fail("'__init__' must be a normal method, not a class or static method!")
-            self.kind = METHOD_INIT
+            if (self.kind is CALLABLE) and cls:
+                self.kind = METHOD_INIT
+            else:
+                fail(
+                    "'__init__' must be a normal method, "
+                    "not a classmethod, staticmethod or getter!"
+                )
```

Comment:
Actually, what do you think about making this error message dynamic? Something like this? (The diff is relative to your PR branch)

```diff
diff --git a/Lib/test/test_clinic.py b/Lib/test/test_clinic.py
index 594bc02105..4641a79ac8 100644
--- a/Lib/test/test_clinic.py
+++ b/Lib/test/test_clinic.py
@@ -638,7 +638,7 @@ class C "void *" ""
             C.__init__ = C.meth
             [clinic start generated code]*/
         """
-        err = "'__init__' must be a normal method, not a classmethod, staticmethod or getter!"
+        err = "'__init__' must be a normal method; got 'FunctionKind.CLASS_METHOD'!"
         self.expect_failure(block, err, lineno=8)

     def test_validate_cloned_new(self):
@@ -2179,18 +2179,23 @@ class Foo "" ""
         """
         self.expect_failure(block, err, lineno=2)

-    def test_init_must_be_a_normal_method(self):
-        err = "'__init__' must be a normal method, not a classmethod, staticmethod or getter!"
-        annotations = ["@classmethod", "@staticmethod", "@getter"]
-        for annotation in annotations:
-            with self.subTest(annotation=annotation):
+    def test_correct_method_kinds_for_special_methods(self):
+        err_template = "'__init__' must be a normal method; got 'FunctionKind.{}'!"
+        annotations = {
+            "@classmethod": "CLASS_METHOD",
+            "@staticmethod": "STATIC_METHOD",
+            "@getter": "GETTER",
+        }
+        for annotation, invalid_kind in annotations.items():
+            with self.subTest(annotation=annotation, invalid_kind=invalid_kind):
                 block = f"""
                     module foo
                     class Foo "" ""
                     {annotation}
                     Foo.__init__
                 """
-                self.expect_failure(block, err, lineno=3)
+                expected_error = err_template.format(invalid_kind)
+                self.expect_failure(block, expected_error, lineno=3)

     def test_duplicate_coexist(self):
         err = "Called @coexist twice"
diff --git a/Tools/clinic/clinic.py b/Tools/clinic/clinic.py
index 9b928ac41a..1f2e5e5adc 100755
--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -5457,8 +5457,8 @@ def update_function_kind(self, fullname: str) -> None:
                 self.kind = METHOD_INIT
             else:
                 fail(
-                    "'__init__' must be a normal method, "
-                    "not a classmethod, staticmethod or getter!"
+                    "'__init__' must be a normal method; "
+                    f"got '{self.kind}'!"
                 )

     def state_modulename_name(self, line: str) -> None:
```

Reply:
Applied!